File: lib/xchain/merkle.go
package xchain

import (
	"github.com/omni-network/omni/lib/errors"
	"github.com/omni-network/omni/lib/merkle"

	"github.com/ethereum/go-ethereum/common"
)

const (
	DSTUnknown merkle.DomainSeparationTag = 0
	DSTHeader  merkle.DomainSeparationTag = 1
	DSTMessage merkle.DomainSeparationTag = 2
)

// MsgTree is a merkle tree of all the messages in a cross-chain block.
// It is used as a leaf when calculating the AttestationRoot.
// Its proofs are used to submit messages to destination chains.
type MsgTree struct {
	tree    [][32]byte       // Merkle tree with all nodes, incl leaves
	indices map[[32]byte]int // Reverse lookup table for node indices
}

func (t MsgTree) MsgRoot() [32]byte {
	return t.tree[0]
}

// Proof returns the merkle multi proof for the provided submissionHeader and messages.
func (t MsgTree) Proof(msgs []Msg) (merkle.MultiProof, error) {
	// Get the indices to prove
	indices := make([]int, 0, len(msgs))
	for _, msg := range msgs {
		msgLeaf, err := msgLeaf(msg)
		if err != nil {
			return merkle.MultiProof{}, err
		}
		msgIndex, err := t.leafIndex(msgLeaf)
		if err != nil {
			return merkle.MultiProof{}, errors.Wrap(err, "msg index")
		}
		indices = append(indices, msgIndex)
	}

	return merkle.GetMultiProof(t.tree, indices...)
}

func (t MsgTree) leafIndex(leaf [32]byte) (int, error) {
	index, ok := t.indices[leaf]
	if !ok {
		return 0, errors.New("leaf not in tree")
	}

	return index, nil
}

// NewMsgTree returns the merkle root of the provided messages
// to be submitted.
func NewMsgTree(msgs []Msg) (MsgTree, error) {
	leafs := make([][32]byte, 0, len(msgs))
	for _, msg := range msgs {
		msgLeaf, err := msgLeaf(msg)
		if err != nil {
			return MsgTree{}, err
		}
		leafs = append(leafs, msgLeaf)
	}

	tree, err := merkle.MakeTree(leafs)
	if err != nil {
		return MsgTree{}, err
	}

	// Create a lookup table for node indices
	indices := make(map[[32]byte]int)
	for i, node := range tree {
		indices[node] = i
	}

	return MsgTree{
		tree:    tree,
		indices: indices,
	}, nil
}

func msgLeaf(msg Msg) ([32]byte, error) {
	bz, err := encodeMsg(msg)
	if err != nil {
		return [32]byte{}, errors.Wrap(err, "encode message")
	}

	return merkle.StdLeafHash(DSTMessage, bz), nil
}

func submissionHeaderLeaf(attHeader AttestHeader, blockHeader BlockHeader) ([32]byte, error) {
	bz, err := encodeSubmissionHeader(attHeader, blockHeader)
	if err != nil {
		return [32]byte{}, errors.Wrap(err, "encode block submissionHeader")
	}

	return merkle.StdLeafHash(DSTHeader, bz), nil
}

// AttestationRoot returns the attestation root of the provided block submissionHeader and message root.
func AttestationRoot(attHeader AttestHeader, blockHeader BlockHeader, msgRoot common.Hash) (common.Hash, error) {
	headerLeaf, err := submissionHeaderLeaf(attHeader, blockHeader)
	if err != nil {
		return [32]byte{}, err
	}

	tree, err := merkle.MakeTree([][32]byte{msgRoot, headerLeaf})
	if err != nil {
		return [32]byte{}, err
	}

	return tree[0], nil
}

File: contracts/core/test/xchain/ts/merkle.ts
/**
 * ts implementation of Omni's XBlock merkle tree. Used for test purposes only,
 * to generate test XSubmissions.
 *
 * Mirrors go implementation in lib/xchain/merkle.go.
 * Uses @openzeppelin/merkle-tree for core merkle tree implementation.
 */

import * as oz from '@openzeppelin/merkle-tree/dist/core'
import { keccak256 } from 'ethereum-cryptography/keccak'
import { equalsBytes, hexToBytes } from 'ethereum-cryptography/utils'
import { Bytes } from '@openzeppelin/merkle-tree/dist/bytes'
import { encodeXMsg, encodeXBlockHeader, XMsg, XBlockHeader } from './xtypes'

type Tree = readonly Bytes[]

type MultiProof = {
  leaves: Bytes[]
  proof: Bytes[]
  proofFlags: boolean[]
}

// XBlockMerkleTree, static methods only
export class XMsgMerkleTree {
  // Make an xblock tree from the provided header and messages
  static of(msgs: readonly XMsg[]): Tree {
    return oz.makeMerkleTree(leafHashes(msgs))
  }

  // Returns a multi proof provided messages, against the provided tree
  static prove(tree: Tree, msgs: readonly XMsg[]): MultiProof {
    const leaves = leafHashes(msgs)
    const indices = leaves.map(l => leafIndex(tree, l))
    return oz.getMultiProof([...tree], indices)
  }

  // Verify a multi proof
  static verify(root: Bytes, p: MultiProof): boolean {
    const impliedRoot = oz.processMultiProof(p)
    return equalsBytes(impliedRoot, root)
  }

  // Get the root of a tree
  static root(tree: Tree): Bytes {
    return tree[0]
  }

  // Order msgs by destination chain id, then offset
  static order(msgs: readonly XMsg[]): readonly XMsg[] {
    // Number(bigint) cast okay, because test chain ids and offsets are small
    return [...msgs].sort((a, b) => {
      if (a.destChainId !== b.destChainId) return Number(a.destChainId - b.destChainId)
      return Number(a.offset - b.offset)
    })
  }
}

const DST_XBLOCK_HEADER = 1
const DST_XMSG = 2

export const attestationRoot = (header: XBlockHeader, msgRoot: Bytes) =>
  oz.makeMerkleTree([msgRoot, headerLeafHash(header)])[0]
const stdLeafHash = (dst: number, data: `0x${string}`) => keccak256(keccak256(new Uint8Array([dst, ...hexToBytes(data)])));

const msgLeafHash = (msg: XMsg) => stdLeafHash(DST_XMSG, encodeXMsg(msg))
const headerLeafHash = (header: XBlockHeader) => stdLeafHash(DST_XBLOCK_HEADER, encodeXBlockHeader(header))

function leafIndex(tree: readonly Bytes[], leaf: Bytes) {
  const index = tree.findIndex(node => equalsBytes(node, leaf))
  if (index === -1) throwError('Leaf is not in tree')
  return index
}

function checkMsgOrder(msgs: readonly XMsg[]) {
  let lastSeenDestChainId = 0n
  let lastSeenOffset = 0n

  for (const msg of msgs) {
    if (msg.destChainId < lastSeenDestChainId) throwError('Msgs not ordered by dest chain id')
    if (msg.destChainId === lastSeenDestChainId && msg.offset < lastSeenOffset)
      throwError('Msgs not ordered by stream offset')

    lastSeenDestChainId = msg.destChainId
    lastSeenOffset = msg.offset
  }
}

function leafHashes(msgs: readonly XMsg[]) {
  checkMsgOrder(msgs)
  return [...msgs.map(msgLeafHash)]
}

function throwError(message?: string) {
  throw new Error(message)
}
